commit 4f2bcda964d4fff56855e0c66198c9bcb682ea1e
Author: Will Newton <will.newton@linaro.org>
Date:   Thu Aug 29 20:10:26 2013 +0100

    ARM: Fix clone code when built for Thumb.
    
    The mov lr, pc instruction will lose the Thumb bit from the return address
    so use blx lr instead.
    
    ports/ChangeLog.arm:
    
    2013-08-30  Will Newton  <will.newton@linaro.org>
    
    	[BZ #15909]
    	* sysdeps/unix/sysv/linux/arm/clone.S (__clone): Use blx
    	instead of mov lr, pc.
    
    (cherry picked from commit 6b06ac56cdfc9293908724e51e827534e97819aa)

---
 NEWS                                      |    6 ++++++
 ports/ChangeLog.arm                       |    6 ++++++
 ports/sysdeps/unix/sysv/linux/arm/clone.S |    4 ++--
 3 files changed, 14 insertions(+), 2 deletions(-)

--- a/NEWS
+++ b/NEWS
@@ -5,6 +5,12 @@
 Please send GNU C library bug reports via <http://sourceware.org/bugzilla/>
 using `glibc' in the "product" field.
 
+Version 2.18.1
+
+* The following bugs are resolved with this release:
+
+  15909.
+
 Version 2.18
 
 * The following bugs are resolved with this release:
--- a/ports/ChangeLog.arm
+++ b/ports/ChangeLog.arm
@@ -1,3 +1,9 @@
+2013-08-30  Will Newton  <will.newton@linaro.org>
+
+	[BZ #15909]
+	* sysdeps/unix/sysv/linux/arm/clone.S (__clone): Use blx
+	instead of mov lr, pc.
+
 2013-07-03  Joseph Myers  <joseph@codesourcery.com>
 
 	* sysdeps/arm/include/bits/setjmp.h [_ISOMAC] (JMP_BUF_REGLIST):
--- a/ports/sysdeps/unix/sysv/linux/arm/clone.S
+++ b/ports/sysdeps/unix/sysv/linux/arm/clone.S
@@ -93,8 +93,8 @@
 	mov	lr, pc
 	bx      ip
 #else
-	mov	lr, pc
-	ldr 	pc, [sp], #8
+	ldr 	lr, [sp], #8
+	blx	lr
 #endif
 
 	@ and we are done, passing the return value through r0
